C error on returning generic anon fn result

### Describe the bug

Running `./v -g -o vdbg cmd/v && ./vdbg 'generic_fn_return.v' && 'generic_fn_return'` on `generic_fn_return.v` produced a compiler error.

### Reproduction Steps

```sh
./v -g -o vdbg cmd/v && ./vdbg 'generic_fn_return.v' && 'generic_fn_return'
```

```v
fn foo[T](do fn () !T) !T {
	mut r := T{}
	r = do()!
	return r
}

fn main() {
	docb := fn () !int {
		return -1
	}
	r := foo(docb)!
	assert r == -1
}
```

### Expected Behavior

Successful compilation and passed assert.

### Current Behavior

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/generic_fn_return.01KR1TN1DGENRFD9V8GM7MN9P3.tmp.c:8553: error: '{' expected (got ";")
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found while compiling generated C code.
This can be caused by invalid C interop code, C compiler flags, or a V compiler bug.
If your code is pure V and this still happens, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

Generated by `v bug` from `generic_fn_return.v`.

### V version

V 0.5.1 701d7f3

### Environment details (OS name and version, etc.)

```
|V full version      |V 0.5.1 2f3c0f1f397c08bfc5ef73e7ff62794471b2b811.701d7f3
|:-------------------|:-------------------
|OS                  |linux, Linux version 7.0.3-arch1-1 (linux@archlinux) (gcc (GCC) 15.2.1 20260209, GNU ld (GNU Binutils) 2.46) #1 SMP PREEMPT_DYNAMIC Thu, 30 Apr 2026 18:41:12 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |4.13GB/15.34GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2026-05-07 18:19:14
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /tmp
|                    |
|Git version         |git version 2.54.0
|V git status        |weekly.2026.03-1869-g701d7f37
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20260209
|gcc version         |gcc (GCC) 15.2.1 20260209
|clang version       |clang version 22.1.3
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43
```

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -1819,7 +1819,11 @@ fn (mut c Checker) infer_fn_generic_types(func &ast.Fn, mut node ast.CallExpr) {
 						}
 						if param_sym.info.func.return_type.has_flag(.generic)
 							&& c.table.sym(param_sym.info.func.return_type).name == gt_name {
-							typ = arg_sym.info.func.return_type.clear_option_and_result()
+							typ = if param_sym.info.func.return_type.has_option_or_result() {
+								arg_sym.info.func.return_type.clear_option_and_result()
+							} else {
+								arg_sym.info.func.return_type
+							}
 							if param_sym.info.func.return_type.nr_muls() > 0 && typ.nr_muls() > 0 {
 								typ = typ.set_nr_muls(0)
 							}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -1819,7 +1819,7 @@ fn (mut c Checker) infer_fn_generic_types(func &ast.Fn, mut node ast.CallExpr) {
 						}
 						if param_sym.info.func.return_type.has_flag(.generic)
 							&& c.table.sym(param_sym.info.func.return_type).name == gt_name {
-							typ = arg_sym.info.func.return_type
+							typ = arg_sym.info.func.return_type.clear_option_and_result()
 							if param_sym.info.func.return_type.nr_muls() > 0 && typ.nr_muls() > 0 {
 								typ = typ.set_nr_muls(0)
 							}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -3890,7 +3890,11 @@ fn (mut g Gen) update_generic_call_concrete_types_from_fn_types(generic_names []
 	}
 	if param_fn.func.return_type.has_flag(.generic) {
 		gt_name := g.table.sym(param_fn.func.return_type).name
-		mut return_type := arg_fn.func.return_type.clear_option_and_result()
+		mut return_type := if param_fn.func.return_type.has_option_or_result() {
+			arg_fn.func.return_type.clear_option_and_result()
+		} else {
+			arg_fn.func.return_type
+		}
 		if arg.expr is ast.LambdaExpr && return_type.has_flag(.generic) {
 			return_type = g.type_resolver.unwrap_generic_expr(arg.expr.expr, return_type)
 			if return_type.has_flag(.generic) {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -3890,7 +3890,7 @@ fn (mut g Gen) update_generic_call_concrete_types_from_fn_types(generic_names []
 	}
 	if param_fn.func.return_type.has_flag(.generic) {
 		gt_name := g.table.sym(param_fn.func.return_type).name
-		mut return_type := arg_fn.func.return_type
+		mut return_type := arg_fn.func.return_type.clear_option_and_result()
 		if arg.expr is ast.LambdaExpr && return_type.has_flag(.generic) {
 			return_type = g.type_resolver.unwrap_generic_expr(arg.expr.expr, return_type)
 			if return_type.has_flag(.generic) {

--- a/vlib/v/tests/generics/generic_fn_callback_result_return_test.v
+++ b/vlib/v/tests/generics/generic_fn_callback_result_return_test.v
@@ -1,13 +1,83 @@
-fn run[T](cb fn () !T) !T {
+fn get_int() int {
+	return 42
+}
+
+fn get_option() ?int {
+	return 42
+}
+
+fn get_result() !int {
+	return 42
+}
+
+fn run_identity[T](cb fn () T) T {
+	return cb()
+}
+
+fn run_option_ret[T](cb fn () T) ?T {
+	mut r := T{}
+	r = cb()
+	return r
+}
+
+fn run_result_ret[T](cb fn () T) !T {
+	mut r := T{}
+	r = cb()
+	return r
+}
+
+fn run_option_cb_option_ret[T](cb fn () ?T) ?T {
+	mut r := T{}
+	r = cb()?
+	return r
+}
+
+fn run_result_cb_option_ret[T](cb fn () !T) ?T {
+	mut r := T{}
+	r = cb() or { panic(err) }
+	return r
+}
+
+fn run_result_cb_result_ret[T](cb fn () !T) !T {
 	mut r := T{}
 	r = cb()!
 	return r
 }
 
-fn test_generic_fn_callback_result_return() {
-	cb := fn () !int {
-		return 42
-	}
+fn test_generic_fn_plain_callback_return_infers_plain_type() {
+	assert run_identity(get_int) == 42
+}
 
-	assert run(cb)! == 42
+fn test_generic_fn_plain_callback_return_infers_option_type() {
+	result := run_identity(get_option)
+	assert result? == 42
+}
+
+fn test_generic_fn_option_callback_return_infers_plain_type() {
+	result := run_option_ret(get_int)
+	assert result? == 42
+}
+
+// TODO: uncomment when issue is fixed (cf. #27876)
+// fn test_generic_fn_option_callback_return_infers_option_type() {
+// 	result := run_option_ret(get_option)
+// 	assert result? == 42
+// }
+
+fn test_generic_fn_result_callback_result_return() {
+	assert run_result_ret(get_int)! == 42
+}
+
+fn test_generic_fn_option_callback_option_return() {
+	result := run_option_cb_option_ret(get_option)
+	assert result? == 42
+}
+
+fn test_generic_fn_result_callback_option_return() {
+	result := run_result_cb_option_ret(get_result)
+	assert result? == 42
+}
+
+fn test_generic_fn_callback_result_return() {
+	assert run_result_cb_result_ret(get_result)! == 42
 }

--- a/vlib/v/tests/generics/generic_fn_callback_result_return_test.v
+++ b/vlib/v/tests/generics/generic_fn_callback_result_return_test.v
@@ -1,0 +1,13 @@
+fn run[T](cb fn () !T) !T {
+	mut r := T{}
+	r = cb()!
+	return r
+}
+
+fn test_generic_fn_callback_result_return() {
+	cb := fn () !int {
+		return 42
+	}
+
+	assert run(cb)! == 42
+}

